include/onnxruntime/core/platform/threadpool.h

```diff
@@ -54,6 +59,43 @@ class ThreadPool {
   // This is not supported until the latest Eigen
   // void SetStealPartitions(const std::vector<std::pair<unsigned, unsigned>>& partitions);
 
+  /**
+  Tries to call the given function in parallel, with calls split into batches of the specified size.
+  **/
+  template <typename F>
+  inline static void TryBatchParallelFor(concurrency::ThreadPool* tp, int32_t total, F&& fn, int32_t batch_size = 0) {
```

Comment:
why &&?

Reply:
R-value with template type makes best effort to hint compilers to inline the function.

Reply:
I consulted VC++ team.
In this case, type F could be:
1. function pointer
2. function object
3. lambda 

In terms of the ability of get inlined, lambdas and function objects works a lot better than function pointer. Function pointer can't get inlined unless it is a compile time const. However, because of some limitations, if fn is a function pointer, you should declare it as F&&(not "const F& fn" or "F fn") to make the constant propagation work. And "F&&" works well for all the three cases with all the popular compilers.

Reply:
F&& is a URef not RRef here, but you std::move it later. If this function is called like below, it will leave f into an unspecified state. 
std::function f;
TryBatchParallelFor(*, f, *);